lib/cfg.js

```diff
@@ -0,0 +1,383 @@
+// SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+// INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+// CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+// ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+// POSSIBILITY OF SUCH DAMAGE.
+
+
+var _ = require('underscore-node'),
+    logger = require('./logger').logger;
+    
+    
+var InstructionType_jmp = 0;
+var InstructionType_conditionalJmpInst = 1;
+var InstructionType_notRetInst = 2;    
+var InstructionType_retInst = 3;    
+
+
+// this mess is intended to be ported to another file.
+function separateCodeFromData(asmArr) {
+    
+    var labelRegex = /\.L{1}\d+:/;
+    
+    var isCode = function(x){
+        return  x && x.text && (x.source !== null || labelRegex.test(x.text) || 
+                !x.text.includes('.'));
+    };
+    
+    return _.chain(asmArr)
+            .map(_.clone)
+            .filter(isCode)
+            .value();
+}
+
+    var gccX86 = {
+                filterData:  separateCodeFromData,
+                isFunctionEnd:function(x)  { return ((x[0] != ' ') && (x[0] != '.') &&
+                                                             (x.indexOf(':') != -1)) ;},
+
+                isBasicBlockEnd: function(x) { return x[0] == ".";} ,
+
+                getInstructionType:function(inst) {
+                                            if(inst.includes("jmp")) return InstructionType_jmp;
+                                            else if(inst.trim()[0] === 'j') return InstructionType_conditionalJmpInst;
+                                            else if(!inst.includes(" ret")) return InstructionType_notRetInst;
+                                            else return InstructionType_retInst;
+                                          },
+
+                extractNodeName:function(inst) {
+                                     var name = inst.match(/\.L{1}\d+/);
+                                     return name + ":";
+                                  },
+
+                isJmpInstruction: function(x) { var trimed = x.trim(); return ( trimed[0] === 'j');}
+             };
+ 
+function separateCodeFromData_(asmArr) {
+    
+    var labelRegex = /\.LBB\d+_\d+:/;
+    var isFunctionName = function(x){
+        var functionNameRegex = /((\w|\(|\))*): # @\1/;
+        return functionNameRegex.test(x);
+    };
+    
+    var isCode = function(x){
+        return  x && x.text && (x.source !== null || labelRegex.test(x.text) || 
+                isFunctionName(x.text));
+    };
+    
+    var removeComments = function(x) {
+        var pos = x.text.indexOf("#");
+        if(pos !== -1)
+            x.text = x.text.substring(0, pos-1);
+        return x;
+    };
+    
+    return _.chain(asmArr)
+            .map(_.clone)
+            .filter(isCode)
+            .map(removeComments)
+            .value();
+}
+
+    var clangX86 = {
```

Comment:
Thanks for adding clang support. I hope we can unify these in future as I hope there's enough in the `gas` syntax common between them. But this is great!

Reply:
Initially I didn't think it would be that easy to add clang support. the two objects are pretty much the same except for code separation, which for clang include comments removal from function name so:
```js
_.chain(asmArr)
            .map(_.clone)
            .filter(isCode)// isCode isn't the same as isCode of gcc
            .map(removeComments)//here
            .value();
```
and jmp labels are denoted differently: .LBB*_*.

so yes it's should be easy enough to find sth that unify the two compilers but my concerns are for different architectures now, I didn't investigate any other architecture so ...

Reply:
@RabsRincon I wasn't aware of monaco tokenization, I am open to your sugestion.

Reply:
@RabsRincon I don't think the Monaco tokeniser is the right route to go this way. It is a great way of expressing syntax colouring, but [a] it's designed as client-side code and [b] it is very overcomplex for most of the parsing we want to do here.

If anything I'd prefer to drive it the other way: have our own ASM parser (that doesn't have to be as hacky and ugly as the Monaco one), and use that to do the syntax highlighting. 

Either way, I think @cppchedy 's approach here is the right decision while we consider our options.